### PR TITLE
EventGraph history

### PR DESCRIPTION
**Pull request by mokaddem**
#### EventGraph export and history

- EventGraph can now be exported in ``json``, ``png``, ``jpeg``, ``dot language``
- Possibility to save network history per ORG
        - only user can delete their network (or the siteAdmin)
        - every user from the same ORG can load the saved eventGraph
- Fixed some bugs

#### Questions
- [X] Does it require a DB change?

#### Release Type:
- [X] Major

### CODE REVIEW

File: app/Controller/EventGraphController.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	public $components = array(
+			'Security',
+			'RequestHandler'
+	);
+
+	public function beforeFilter() {
+		parent::beforeFilter();
+	}
+
+	public function get($event_id = false) {
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Could we change this to view? ;)

File: app/Controller/EventGraphController.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		if (!$this->request->is('get')) {
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Generally there's no need to block POST requests for GET actions, only the other way around. Consider that you might want to inject some parameters at some point via a JSON payload

File: app/Controller/EventGraphController.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			throw new MethodNotAllowedException(__('Invalid method.'));
+		}
+		if ($event_id === false) throw new MethodNotAllowedException(__('No event ID set.'));
+
+		// retreive current org_id
+		$org_id = $this->_checkOrg();
+
+		// validate event
+		$this->loadModel('Event');
+		if (!is_numeric($event_id)) {
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Perhaps also allow using UUIDs in addition to event IDs. Always consider the API use-case - do I ever want to be able to fetch the graph JSON's via the API?

File: app/Controller/EventGraphController.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		// fetch eventGraphs
+		$eventGraphs = $this->EventGraph->find('all', array(
+			'order' => 'EventGraph.timestamp DESC',
+			'conditions' => array(
+				'EventGraph.event_id' => $event_id,
+				'EventGraph.org_id' => $org_id
+			),
+			'contain' => array(
+				'User' => array(
+					'conditions' => array(
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: No need for this if the correct model linking is set up.

File: app/Controller/EventGraphController.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+				)
+			)
+		));
+		return $this->RestResponse->viewData($eventGraphs, $this->response->type());
+	}
+
+	public function add($event_id = false) {
+		if ($this->request->is('get') && $this->_isRest()) {
+			return $this->RestResponse->describe('EventGraph', 'add', false, $this->response->type());
+		} else if ($this->request->is('get')) { // retreive form
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Perhaps a structure like would be a bit nicer:

```
if ($this->request->is('get')) {
    if ($this->_isRest()) {
        #describe API and return
    }
    # retrieve form
} else {
    #do post action
}

File: app/Controller/EventGraphController.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+					throw new NotFoundException(__('Invalid event'));
+				}
+			}
+
+			$this->set('action', 'add');
+			$this->set('event_id', $event_id);
+			$this->render('ajax/' . $formURL);
+
+		} else {
+			if ($event_id === false) throw new MethodNotAllowedException(__('No event ID set.'));
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: if (!empty($event_id))

File: app/Controller/EventGraphController.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+			$this->loadModel('Event');
+			$event = $this->Event->fetchEvent($this->Auth->user(), array('eventid' => $event_id));
+			if (empty($event)) throw new NotFoundException('Invalid event');
+
+			$eventGraph = array();
+			if (!$this->_isSiteAdmin() && ($event['Event']['orgc_id'] != $this->_checkOrg() && !$this->userRole['perm_modify'])) {
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: $this->Auth->user('org_id')

File: app/Controller/EventGraphController.php [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+				$eventGraph['EventGraph']['network_name'] = null;
+			} else {
+				$eventGraph['EventGraph']['network_name'] = $this->request->data['EventGraph']['network_name'];
+			}
+
+			if (isset($this->request->data['EventGraph']['preview_img'])) {
+				$eventGraph['EventGraph']['preview_img'] = $this->request->data['EventGraph']['preview_img'];
+			}
+
+			$eventGraph['EventGraph']['timestamp'] = $date->getTimestamp();
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Move this to beforeValidate() in the model

File: app/Controller/EventGraphController.php [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+			// Network pushed will be the owner of the authentication key
+			$eventGraph['EventGraph']['user_id'] = $this->Auth->user('id');
+			$eventGraph['EventGraph']['org_id'] = $this->Auth->user('org_id');
+
```
Comment (iglocska) [inline app/Controller/EventGraphController.php:99]: $this->EventGraph->create();

File: app/Controller/EventGraphController.php [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+		}
+	}
+
+	public function delete($id) {
+		if (!$this->request->is('post')) {
+			$this->set('id', $id);
+			$this->render('ajax/eventGraph_delete_form');
+		} else {
+			$this->set('id', $id);
+			$conditions = array('id' => $id);
```
Comment (iglocska) [inline app/Controller/EventGraphController.php:124]: Add to the conditions org_id == user's org_id
```
$conditions = array(
   'id' => $id
);
if (!$this->_isSiteAdmin()) {
    $conditions['org_id'] = $this->Auth->user('org_id');
}
```

File: app/Controller/EventGraphController.php [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+			$eventGraph = $this->EventGraph->find('first', array(
+					'conditions' => $conditions,
+					'recursive' => -1,
+					'fields' => array('id', 'event_id', 'user_id'),
+			));
+			if (empty($eventGraph)) throw new NotFoundException('Invalid EventGraph');
+			if ($this->request->is('ajax')) {
```
Comment (iglocska) [inline app/Controller/EventGraphController.php]: Not needed - API users should be able to delete them too

File: app/Model/AppModel.php [hunk 12]
```diff
@@ -958,9 +958,25 @@ public function updateDatabase($command) {
...
+					`user_id` int(11) NOT NULL,
+					`org_id` int(11) NOT NULL,
+					`timestamp` int(11) NOT NULL DEFAULT 0,
+					`network_name` varchar(255) CHARACTER SET utf8 COLLATE utf8_unicode_ci,
+					`network_json` MEDIUMTEXT NOT NULL,
+					`preview_img` MEDIUMTEXT,
+					PRIMARY KEY (id),
+					INDEX `event_id` (`event_id`),
+					INDEX `user_id` (`user_id`),
+					INDEX `org_id` (`org_id`)
```
Comment (iglocska) [inline app/Model/AppModel.php:979]:
```
INDEX `timestamp` (`timestamp`)
```

File: app/Model/EventGraph.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+			'foreignKey' => 'user_id',
+			'conditions' => '',
+			'fields' => '',
+			'order' => ''
+		)
+	);
+
+
+	public $validate = array(
+		'is_json' => array(
```
Comment (iglocska) [inline app/Model/EventGraph.php]: Add the field name

File: app/Model/EventGraph.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+		),
+	);
+
+	public function beforeValidate($options = array()) {
+		parent::beforeValidate();
+		return true;
+	}
+
+	public function isValidJson($text) {
+		$check = json_decode($text);
```
Comment (iglocska) [inline app/Model/EventGraph.php]: try catch (maybe, not sure)

File: app/Controller/EventGraphController.php (hunk 5 above)
Comment (mokaddem) [reply app/Controller/EventGraphController.php]:
```
/Perhaps//
/would be a bit/is/
```

File: app/Model/EventGraph.php (hunk 14 above)
Comment (mokaddem) [reply app/Model/EventGraph.php]: Apparently, this is not a true validation... see 
https://stackoverflow.com/questions/7841670/validate-json-in-php/47227564

Do I keep it like this?

**Comment by mokaddem** [comment]
PR updated, feel free to go through once again ;)

**Comment by iglocska** [comment]
Thanks a lot!

**Pull request merged**